### PR TITLE
Add new practice exercise `zebra-puzzle`

### PR DESCRIPTION
**Pull request by jiegillet**
I felt like adding this one. It's a hard problem and I was hoping the use of types might help building a clear solution, but in the end, not that much...

In my example implementation I used generics (for example in `sameHouseConstraint`), so we could say it practices generics, but I'm not really sure it's the only approach.

### CODE REVIEW

File: exercises/practice/zebra-puzzle/.meta/src/ZebraPuzzle.example.elm [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+            List.all (\h -> field h /= field house || .number h == .number house) houses
+    in
+    if houseOnlyHasOneTypeOfField || typeOfFieldIsOnlyFoundInHouse then
+        Just (sameHouseConstraint field (field house) .number (.number house))
+
+    else
+        Nothing
+
+
+unique : List a -> List a
```
Comment (ceddlyburge) [inline exercises/practice/zebra-puzzle/.meta/src/ZebraPuzzle.example.elm:314]: I think using a Set here is more usual in Elm

File: exercises/practice/zebra-puzzle/.meta/src/ZebraPuzzle.example.elm [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+                possibilities
+
+            else
+                iterate newPossibilities
+    in
+    iterate reducedPossibilities
+
+
+allPossibilities : List House
+allPossibilities =
```
Comment (ceddlyburge) [inline exercises/practice/zebra-puzzle/.meta/src/ZebraPuzzle.example.elm:99]: Is there something in List.Extra that makes this easier?

File: exercises/practice/zebra-puzzle/tests/Tests.elm [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+import Expect
+import Test exposing (Test, describe, skip, test)
+import ZebraPuzzle
+
+
+tests : Test
+tests =
+    describe "ZebraPuzzle"
+        [ test "resident who drinks water" <|
```
Comment (ceddlyburge) [inline exercises/practice/zebra-puzzle/tests/Tests.elm:11]: This code / tests are a bit strange, as the answer is already known, so it would be very easy to pass the tests just by returning these values, which is what you would do if you were using tdd. This is a standard exercise though, right? So probably not something we should worry about here.

**Review by ceddlyburge — APPROVED** [review]

File: exercises/practice/zebra-puzzle/.meta/src/ZebraPuzzle.example.elm (hunk 1 above)
Comment (jiegillet) [reply exercises/practice/zebra-puzzle/.meta/src/ZebraPuzzle.example.elm:314]: Of course, but `Set` won't take a custom types, and I didn't feel like defining some `toComparable` and `fromComparable` for all 5 types of custom types I use this function on. I really wish `Set` and `Dict` could handle any type out of the box.

File: exercises/practice/zebra-puzzle/.meta/src/ZebraPuzzle.example.elm (hunk 2 above)
Comment (jiegillet) [reply exercises/practice/zebra-puzzle/.meta/src/ZebraPuzzle.example.elm:99]: Mmh, yeah maybe `cartesianProduct` + `transpose` + `List.map5` could make it cleaner. `unique` is also in there.
But we don't have access to `List.Extra` with the standard `elm.json`.

File: exercises/practice/zebra-puzzle/tests/Tests.elm (hunk 3 above)
Comment (jiegillet) [reply exercises/practice/zebra-puzzle/tests/Tests.elm:11]: Right, this is a standard exercise, and incredibly easy to pass if you cheat.
I was thinking of adding an analyzer check to make sure that the functions are not exactly `drinksWater = Just Norwegian`, but beyond that, I'm not too worried about it.

File: exercises/practice/zebra-puzzle/.meta/src/ZebraPuzzle.example.elm (hunk 2 above)
Comment (ceddlyburge) [reply exercises/practice/zebra-puzzle/.meta/src/ZebraPuzzle.example.elm:99]: If its a lot easier, then its probably worth a comment at least, so that a student will know there is a better way, even if we don't do it in the exemplar.

Comment (jiegillet) [reply exercises/practice/zebra-puzzle/.meta/src/ZebraPuzzle.example.elm:99]: I don't think it's that much easier, even with the functions I suggested you would need a messy pattern match for a list of 5 elements... And using another package will mean the test-runner and the analyzer will not run, so I think it's a suggestion with potential negative consequences. 
 
The way I did it looks messy but it's not complex, it's the same thing repeated 5 times. It's probably fine.
